File: Exercise_2.cpp
#include <iostream>
#include <stdio.h>
#include <stdlib.h>

using namespace std; 

#define EMPTY -9999

// A structure to represent a stack
class StackNode {

public:
    int data;
    StackNode* next;
   
    StackNode() {
        next = NULL;
    }
};

StackNode* top = NULL;

StackNode* newNode(int data)
{
    StackNode* stackNode = new StackNode();
    stackNode->data = data;
    stackNode->next = NULL;
    return stackNode;
}
 
bool isEmpty(StackNode* root)
{
    if (NULL == top)
        return true;
    else
        return false;
}
 
void push(StackNode* root, int data)
{
    if (NULL == root)
        return;
   
    if (NULL == top) {
        top = root;
        top->data = data;
        return;
    }
   
    top->next = newNode(data);
    top = top->next;
   
}
 
int pop(StackNode* root)
{
    if (NULL == root)
        return EMPTY;
    if (NULL == top)
        return EMPTY;
   
    if (root == top) {
        int data = top->data;
        top = NULL;
        return data;
    }  
    int data = top->data;
   
    StackNode* cur = root ;
    while (cur->next != top) {
        cur = cur->next;
    }
    cur->next = NULL;
    top = cur;
   
   
    return data;
}
 
int peek(StackNode* root)
{
    if (NULL == root)
        return EMPTY;
    if (NULL == top)
        return EMPTY;
       
    return top->data;
}

int main()
{
    StackNode* root = new StackNode();
   
 
    push(root, 10);
    push(root, 20);
    push(root, 30);
 
    cout << pop(root) << " popped from stack\n";
 
    cout << "Top element is " << peek(root) << endl;
 
    return 0;
} 

File: Exercise2.cs
using System;
using System.Security.Cryptography;

namespace Exercise2
{
    class StackLL
    {
        Node top;
        public StackLL()
        {
            this.top = null;
        }
        public void push(int x)
        {
            Node temp = new Node();
            if (temp!=null)
            {
                temp.data = x;
                temp.linknode = top;
                top = temp;
            }
        }
        public bool isempty()
        {
            if(top==null)
            { return true; }
            else return false;
        }

        public int peek()
        {
            if (!isempty())
            {
                return top.data;
            }
            else
                return -1;
        }

        public int pop()
        {
            if(top==null)
            {
                return 0;
            }
            else
            {
                int i = top.data;
                top =top.linknode;
                return i;
                
            }
        }
        static void Main(string[] args)
        {
            StackLL sl = new StackLL();
            sl.push(10);
            sl.push(20);
            sl.push(30);
            Console.WriteLine("following element popped from stack {0}", sl.pop());
            Console.WriteLine("top element is {0}", sl.peek());
        }
    }
    class Node
    {
        public int data;
        public Node linknode;
    }
}
